cpp/src/arrow/compute/kernels/filter.cc

```diff
@@ -15,44 +15,453 @@
+                std::shared_ptr<Array>* out) override {
+    out->reset(new NullArray(OutputSize(checked_cast<const BooleanArray&>(filter))));
+    return Status::OK();
+  }
+};
+
+template <typename ValueType>
+class FilterImpl : public FilterKernel {
+ public:
+  using ValueArray = typename TypeTraits<ValueType>::ArrayType;
+  using OutBuilder = typename TypeTraits<ValueType>::BuilderType;
+
+  using FilterKernel::FilterKernel;
+
+  Status Filter(FunctionContext* ctx, const Array& values, const BooleanArray& filter,
+                std::shared_ptr<Array>* out) override {
+    std::unique_ptr<OutBuilder> builder;
+    RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), type_, &builder));
+    RETURN_NOT_OK(builder->Resize(OutputSize(filter)));
+    RETURN_NOT_OK(UnpackValuesNullCount(checked_cast<const ValueArray&>(values), filter,
+                                        builder.get()));
+    return builder->Finish(out);
+  }
+
+ private:
+  Status UnpackValuesNullCount(const ValueArray& values, const BooleanArray& filter,
+                               OutBuilder* builder) {
+    if (values.null_count() == 0) {
+      return UnpackIndicesNullCount<true>(values, filter, builder);
+    }
+    return UnpackIndicesNullCount<false>(values, filter, builder);
+  }
+
+  template <bool AllValuesValid>
+  Status UnpackIndicesNullCount(const ValueArray& values, const BooleanArray& filter,
+                                OutBuilder* builder) {
+    if (filter.null_count() == 0) {
+      return Filter<AllValuesValid, true>(values, filter, builder);
+    }
+    return Filter<AllValuesValid, false>(values, filter, builder);
+  }
+
+  template <bool AllValuesValid, bool AllIndicesValid>
+  Status Filter(const ValueArray& values, const BooleanArray& filter,
+                OutBuilder* builder) {
+    for (int64_t i = 0; i < filter.length(); ++i) {
+      if (!AllIndicesValid && filter.IsNull(i)) {
+        builder->UnsafeAppendNull();
+        continue;
+      }
+      if (!filter.Value(i)) {
+        continue;
+      }
+      if (!AllValuesValid && values.IsNull(i)) {
+        builder->UnsafeAppendNull();
+        continue;
+      }
+      RETURN_NOT_OK(UnsafeAppend(builder, values.GetView(i)));
+    }
+    return Status::OK();
+  }
+};
+
+template <>
+class FilterImpl<StructType> : public FilterKernel {
+ public:
+  using FilterKernel::FilterKernel;
+
+  Status Filter(FunctionContext* ctx, const Array& values, const BooleanArray& filter,
+                std::shared_ptr<Array>* out) override {
+    const auto& struct_array = checked_cast<const StructArray&>(values);
+
+    auto length = OutputSize(filter);
+    TypedBufferBuilder<bool> null_bitmap_builder(ctx->memory_pool());
+    RETURN_NOT_OK(null_bitmap_builder.Resize(length));
+
+    ArrayVector fields(type_->num_children());
+    for (int i = 0; i < type_->num_children(); ++i) {
+      RETURN_NOT_OK(
+          arrow::compute::Filter(ctx, *struct_array.field(i), filter, &fields[i]));
```

Comment:
In the other nested data, you opted for storing child' kernels (via impl_ property) instead of explicit recurse. Is there a reason for the divergence? This is a not a blocker, just curious.

This would fix the concern of re-computing OutputSize.

Reply:
FilterImpl for struct was the first impl I wrote for nested types, and it was easiest to just recurse. However this should probably also hold child kernels for later on when the kernel may be reused. This will not fix recomputing OutputSize however; I'll need to add a hint function so that FilterImpl<StructType> can pass OutputSize to its children